RFC: Omnilock (#343)
nervosnetwork/rfcs: rfcs/0042-omnilock/0042-omnilock.md

---
Number: "0042"
Category: Standards Track
Status: Proposal
Author: Xu Jiandong <lynndon@gmail.com>
Created: 2022-05-19
---

# Omnilock

> Omnilock is a new lock script designed for interoperability. It supports various transaction verifications used in

Reviewer: ```suggestion
Omnilock is a lock script designed for interoperability. It comes with built-in support for verification of transaction signing methods used in
```

popular blockchains, such as Bitcoin, Ethereum, EOS, and Dogecoin. Omnilock is also extensible, so more verification

> algorithms can be added in future. This feature makes Omnilock a powerful module in Nervos interoperability 2.0.

Reviewer: ```suggestion
algorithms can be added in future.
```

> Another feature of Omnilock for practitioners is the regulation compliance module (Consider it as interoperability with

Reviewer: ```suggestion
Another feature of Omnilock for practitioners is the regulation compliance module which brings interoperability with
```

the traditional world). If enabled, the specified administrator can revoke tokens held by users under circumstances
which the administrator deems proper. This part has evolved from the [Regulation Compliance Extension
(RCE)](https://talk.nervos.org/t/rfc-regulation-compliance-extension/5338) proposal for
[xUDT](https://talk.nervos.org/t/rfc-extensible-udt/5337). This feature provides an option that sits at the other side
of the asset lock spectrum and lays the foundation of registered assets like Apple stock on CKB. When used together,
Omnilock and RCE provide an [ERC-1404](https://erc1404.org/) equivalence.

## Authentication

> Omnilock introduces a new concept, authentication ( auth ) to CKB lock scripts: an auth is a 21-byte data structure

Reviewer: ```suggestion
An authentication (auth) is a 21-byte data structure
```

```

| Name               | Flags      | Affected Args   |Affected Args Size (byte)|Affected Witness
| -------------------|------------|-----------------|---------|-------------------------------- 
| administrator mode | 0b00000001 |	AdminList cell Type ID | 32      | omni_identity/signature in OmniLockWitnessLock
| anyone-can-pay mode| 0b00000010 | minimum ckb/udt in ACP| 2 | N/A
| time-lock mode     | 0b00000100 | since for timelock| 8     | N/A
| supply mode        | 0b00001000 |type script hash for supply| 32 | N/A
| auth               | N/A        |21-byte auth identity | 21 | signature in OmniLockWitnessLock

## Administrator Mode

When "administrator mode" is enabled, `<32 byte AdminList cell Type ID>` must be present. The AdminList cell contains the
type script hash used by a special cell with the same format as [RCE
Cell](https://talk.nervos.org/t/rfc-regulation-compliance-extension/5338). The RCE cell follows a set of rules and contains
whitelists and blacklists. These lists can be used in the [SMT proofs
scenarios](https://github.com/nervosnetwork/sparse-merkle-tree).

The RCE cells are organized in tree structure illustrated in the following diagram:

> ![RCE Cells](./rce_cells.png)

Reviewer 2: Please share the diagram source file in comment.

Author: https://drive.google.com/file/d/1mKgqIl2IK-n9JV0kktT2zZDvDaRsUS7D/view?usp=sharing
Open it with draw.io

    preimage: BytesOpt,
}
```

When `omni_identity` is present, it will be validated whether the provided auth in `omni_identity` is present in RC
AdminList Cell associated with the current lock script via SMT validation rules. In this case, the auth included in
`omni_identity` will be used in further validation.

If `omni_identity` is missing, the auth included in lock script args will then be used in further validation.

Once the processing above is successfully done and the auth to be used is confirmed, the flag in the designated auth
will be checked for the succeeding operations:

* When the auth flag is 0x0, a signature must be present in `OmniLockWitnessLock`. We will use the signature for secp256k1
  recoverable signature verification. The recovered public key hash using the blake160 algorithm must match the current
  auth content.

* When the auth flag is 0xFC, we will check against the current transaction, and there must be an input cell, whose lock
  script matches the auth content when hashed via blake160.

When `signature` is present, the signature can be used to unlock the cell in anyone-can-pay mode.

When `preimage` is present, the auth flag can be 0xFD or 0xFE.

> * When auth flag is 0xFD, the `exec` method is used. The preimage's memory layout will be as follows:

Reviewer 3: suggest to add a prefix to preimage layout, it can be used as a tag or extension bit:

| Flag | Preimage Layout | Verification |
| ------------- | ------------- | ------------- |
| 0x01  | exec code hash(32 bytes) exec hash type (1 byte)...  | lock.args[0] == 0xFD && lock.args[1..21] == blake2b(preimage)|
| 0x02  | dynamic library code hash (32 bytes) ...| lock.args[0] == 0xFE && lock.args[1..21] = blake2b(preimage)|

```
exec code hash (32 bytes)
exec hash type (1 byte)
place (1 byte)
bounds (8 bytes)
pubkey blake160 hash (20 bytes)
```

The `place` and `bounds` are passed in directly to [exec](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0034-vm-syscalls-2/0034-vm-syscalls-2.md#exec):
```C

> int ckb_exec(size_t index, size_t source, size_t place, size_t bounds, int argc, char* argv[]);

Reviewer 3: Since the source can only be cell deps, why do we need to set the place parameter? It is only legal if it is 0, suggest to remove the `place (1 byte)` from the preimage

Author: This change requires the implementation to be updated. But the implementation has been deployed.
If we have further version of omnilock, I think it's OK to accept this change.

```
The `index` is located by `exec code hash` and `exec hash type` from `cell_deps`([See usage](https://github.com/nervosnetwork/ckb-c-stdlib/blob/20578dfb092b3b3761df755395e20ec142a83d6e/ckb_syscalls.h#L368)].

> Finally, message, signature, pubkey blake160 hash are encoded into hex strings. Then these hex strings are passed in as `argv`.

Reviewer 3: What does the `message` here mean specifically? Can you add more descriptions?

* When auth flag is 0xFE, the `dynamic linking` method is used. The preimage's memory layout will be as follows:
```
dynamic library code hash (32 bytes)
dynamic library hash type (1 byte)
pubkey blake160 hash （20 bytes)
```

It loads the dynamic linking libraries via `code hash` and `hash type`, and gets the entry function named 
`validate_signature`. The entry function is expected to have following C API:
```C
int validate_signature(void *prefilled_data, const uint8_t *signature_buffer,
    size_t signature_size, const uint8_t *message_buffer, size_t message_size,
    uint8_t *pubkey_hash, size_t *pubkey_hash_len);
```

> Then the entry function is called to validate the message and signature. The `pubkey_hash` returned from the entry

Reviewer 3: same question, what does the message here mean specifically?

Author: The description of the `message` is added in the section `Omnilock Witness`.
